File: packages/sdk-go/types/module.go
package types

import (
	"fmt"
	"math/big"
)

// ModuleType represents the ERC-7579 module type.
type ModuleType uint8

// ERC-7579 Module Types
const (
	ModuleTypeValidator ModuleType = 1
	ModuleTypeExecutor  ModuleType = 2
	ModuleTypeFallback  ModuleType = 3
	ModuleTypeHook      ModuleType = 4
	ModuleTypePolicy    ModuleType = 5
	ModuleTypeSigner    ModuleType = 6
)

// String returns the string name of the module type.
func (m ModuleType) String() string {
	switch m {
	case ModuleTypeValidator:
		return "Validator"
	case ModuleTypeExecutor:
		return "Executor"
	case ModuleTypeFallback:
		return "Fallback"
	case ModuleTypeHook:
		return "Hook"
	case ModuleTypePolicy:
		return "Policy"
	case ModuleTypeSigner:
		return "Signer"
	default:
		return "Unknown"
	}
}

// ModuleStatus represents the installation status of a module.
type ModuleStatus uint8

const (
	ModuleStatusNotInstalled ModuleStatus = iota
	ModuleStatusInstalling
	ModuleStatusInstalled
	ModuleStatusUninstalling
	ModuleStatusFailed
)

// String returns the string name of the module status.
func (s ModuleStatus) String() string {
	switch s {
	case ModuleStatusNotInstalled:
		return "not_installed"
	case ModuleStatusInstalling:
		return "installing"
	case ModuleStatusInstalled:
		return "installed"
	case ModuleStatusUninstalling:
		return "uninstalling"
	case ModuleStatusFailed:
		return "failed"
	default:
		return "unknown"
	}
}

// MarshalJSON serializes ModuleStatus as a JSON string for API compatibility with sdk-ts.
func (s ModuleStatus) MarshalJSON() ([]byte, error) {
	return []byte(`"` + s.String() + `"`), nil
}

// UnmarshalJSON deserializes ModuleStatus from a JSON string.
func (s *ModuleStatus) UnmarshalJSON(data []byte) error {
	str := string(data)
	// Strip quotes
	if len(str) >= 2 && str[0] == '"' && str[len(str)-1] == '"' {
		str = str[1 : len(str)-1]
	}
	switch str {
	case "not_installed":
		*s = ModuleStatusNotInstalled
	case "installing":
		*s = ModuleStatusInstalling
	case "installed":
		*s = ModuleStatusInstalled
	case "uninstalling":
		*s = ModuleStatusUninstalling
	case "failed":
		*s = ModuleStatusFailed
	default:
		return fmt.Errorf("unknown ModuleStatus: %s", str)
	}
	return nil
}

// InstalledModule represents an installed module on a smart account.
type InstalledModule struct {
	Type       ModuleType   `json:"type"`
	Address    Address      `json:"address"`
	InitData   Hex          `json:"initData,omitempty"`
	Status     ModuleStatus `json:"status"`
	InstalledAt *big.Int    `json:"installedAt,omitempty"`
}

// ModuleMetadata provides metadata about a module.
type ModuleMetadata struct {
	Name        string     `json:"name"`
	Description string     `json:"description"`
	Version     string     `json:"version"`
	Author      string     `json:"author,omitempty"`
	Type        ModuleType `json:"type"`
	Tags        []string   `json:"tags,omitempty"`
	IsAudited   bool       `json:"isAudited"`
	AuditUrl    string     `json:"auditUrl,omitempty"`
}

// ModuleInstallRequest represents a request to install a module.
type ModuleInstallRequest struct {
	Type     ModuleType `json:"type"`
	Address  Address    `json:"address"`
	InitData Hex        `json:"initData,omitempty"`
}

// ModuleUninstallRequest represents a request to uninstall a module.
type ModuleUninstallRequest struct {
	Type       ModuleType `json:"type"`
	Address    Address    `json:"address"`
	DeInitData Hex        `json:"deInitData,omitempty"`
}

// ModuleForceUninstallRequest represents a request to force-uninstall a module.
// Uses ExcessivelySafeCall - module revert is ignored.
type ModuleForceUninstallRequest struct {
	Type       ModuleType `json:"type"`
	Address    Address    `json:"address"`
	DeInitData Hex        `json:"deInitData,omitempty"`
}

// ModuleReplaceRequest represents a request to atomically replace a module.
// Supported for VALIDATOR, EXECUTOR, FALLBACK types.
type ModuleReplaceRequest struct {
	Type       ModuleType `json:"type"`
	OldAddress Address    `json:"oldAddress"`
	DeInitData Hex        `json:"deInitData,omitempty"`
	NewAddress Address    `json:"newAddress"`
	InitData   Hex        `json:"initData,omitempty"`
}

// HookGasLimitRequest represents a request to set per-hook gas limit.
// GasLimit of 0 means unlimited (backward compatible).
type HookGasLimitRequest struct {
	HookAddress Address  `json:"hookAddress"`
	GasLimit    *big.Int `json:"gasLimit"`
}

// DelegatecallWhitelistRequest represents a request to set delegatecall whitelist entry.
type DelegatecallWhitelistRequest struct {
	Target  Address `json:"target"`
	Allowed bool    `json:"allowed"`
}

// DelegatecallWhitelistEnforceRequest represents a request to enable/disable delegatecall whitelist enforcement.
type DelegatecallWhitelistEnforceRequest struct {
	Enforce bool `json:"enforce"`
}

// Module operation error constants matching Kernel.sol errors.
const (
	ErrDelegatecallTargetNotWhitelisted = "DelegatecallTargetNotWhitelisted"
	ErrReentrancy                       = "Reentrancy"
)

// ECDSAValidatorConfig represents ECDSA validator initialization config.
type ECDSAValidatorConfig struct {
	Owner Address `json:"owner"`
}

// WebAuthnValidatorConfig represents WebAuthn validator initialization config.
type WebAuthnValidatorConfig struct {
	PubKeyX      *big.Int `json:"pubKeyX"`
	PubKeyY      *big.Int `json:"pubKeyY"`
	CredentialId Hex      `json:"credentialId"`
}

// MultiSigValidatorConfig represents multi-sig validator initialization config.
type MultiSigValidatorConfig struct {
	Signers   []Address `json:"signers"`
	Threshold uint8     `json:"threshold"`
}

// SessionKeyConfig represents session key executor initialization config.
type SessionKeyConfig struct {
	SessionKey      Address   `json:"sessionKey"`
	AllowedTargets  []Address `json:"allowedTargets"`
	AllowedSelectors []Hex    `json:"allowedSelectors,omitempty"`
	SpendLimit      *big.Int  `json:"spendLimit,omitempty"`
	ValidAfter      uint64    `json:"validAfter"`
	ValidUntil      uint64    `json:"validUntil"`
}

// RecurringPaymentConfig represents recurring payment executor config.
type RecurringPaymentConfig struct {
	Recipient     Address  `json:"recipient"`
	Token         Address  `json:"token"`
	Amount        *big.Int `json:"amount"`
	Interval      uint64   `json:"interval"` // seconds
	MaxExecutions uint64   `json:"maxExecutions"`
}

// SpendingLimitConfig represents spending limit hook config.
type SpendingLimitConfig struct {
	Token     Address  `json:"token"`
	Limit     *big.Int `json:"limit"`
	Period    uint64   `json:"period"` // seconds
	Recipient *Address `json:"recipient,omitempty"`
}

// AuditHookConfig represents audit hook configuration.
type AuditHookConfig struct {
	AuditAddress Address `json:"auditAddress"`
	EventFlags   uint32  `json:"eventFlags"`
}

// TokenReceiverConfig represents token receiver fallback configuration.
type TokenReceiverConfig struct {
	SupportedInterfaces []Hex `json:"supportedInterfaces"`
}

// FlashLoanConfig represents flash loan fallback configuration.
type FlashLoanConfig struct {
	AuthorizedBorrower Address            `json:"authorizedBorrower"`
	AllowedTokens      []Address          `json:"allowedTokens"`
	MaxLoanAmounts     map[Address]*big.Int `json:"maxLoanAmounts,omitempty"`
}

File: packages/types/src/module.ts
import type { Address, Hex } from 'viem'

// ============================================================================
// Constants
// ============================================================================

/**
 * ERC-7579 Module Types
 * @see https://eips.ethereum.org/EIPS/eip-7579
 */
export const MODULE_TYPE = {
  /** Validates transaction signatures */
  VALIDATOR: 1n,
  /** Executes specific transaction logic */
  EXECUTOR: 2n,
  /** Handles unsupported function calls */
  FALLBACK: 3n,
  /** Pre/post transaction execution hooks */
  HOOK: 4n,
  /** Policy validation for execution */
  POLICY: 5n,
  /** Multi-signature participant */
  SIGNER: 6n,
} as const

export type ModuleType = (typeof MODULE_TYPE)[keyof typeof MODULE_TYPE]

/**
 * Module type names for display
 */
export const MODULE_TYPE_NAMES: Map<bigint, string> = new Map([
  [MODULE_TYPE.VALIDATOR, 'Validator'],
  [MODULE_TYPE.EXECUTOR, 'Executor'],
  [MODULE_TYPE.FALLBACK, 'Fallback'],
  [MODULE_TYPE.HOOK, 'Hook'],
  [MODULE_TYPE.POLICY, 'Policy'],
  [MODULE_TYPE.SIGNER, 'Signer'],
])

/**
 * Module installation status
 */
export const MODULE_STATUS = {
  /** Not installed */
  NOT_INSTALLED: 'not_installed',
  /** Installation in progress (tx submitted) */
  INSTALLING: 'installing',
  /** Installed and active */
  INSTALLED: 'installed',
  /** Uninstallation in progress (tx submitted) */
  UNINSTALLING: 'uninstalling',
  /** Installation or operation failed */
  FAILED: 'failed',
} as const

export type ModuleStatus = (typeof MODULE_STATUS)[keyof typeof MODULE_STATUS]

// ============================================================================
// Interfaces
// ============================================================================

/**
 * Module metadata from registry
 */
export interface ModuleMetadata {
  /** Module contract address */
  address: Address

  /** Module type */
  type: ModuleType

  /** Human-readable name */
  name: string

  /** Description */
  description: string

  /** Version string */
  version: string

  /** Author/maintainer */
  author?: string

  /** Documentation URL */
  docsUrl?: string

  /** Source code URL */
  sourceUrl?: string

  /** Security audit URL */
  auditUrl?: string

  /** Is this an official/verified module */
  isVerified: boolean

  /** Tags for filtering */
  tags: string[]

  /** Logo URL */
  logoUrl?: string
}

/**
 * Module configuration schema
 * Defines what parameters the module accepts during installation
 */
export interface ModuleConfigSchema {
  /** Schema version */
  version: string

  /** Configuration fields */
  fields: ModuleConfigField[]
}

/**
 * Single configuration field
 */
export interface ModuleConfigField {
  /** Field name (used in encoding) */
  name: string

  /** Display label */
  label: string

  /** Field description */
  description: string

  /** Solidity type (address, uint256, bytes, etc.) */
  type: SolidityType

  /** Is this field required */
  required: boolean

  /** Default value (if any) */
  defaultValue?: string

  /** Validation rules */
  validation?: FieldValidation
}

/**
 * Solidity types for encoding
 */
export type SolidityType =
  | 'address'
  | 'uint8'
  | 'uint16'
  | 'uint32'
  | 'uint64'
  | 'uint128'
  | 'uint256'
  | 'int8'
  | 'int16'
  | 'int32'
  | 'int64'
  | 'int128'
  | 'int256'
  | 'bool'
  | 'bytes'
  | 'bytes4'
  | 'bytes32'
  | 'string'
  | 'address[]'
  | 'uint256[]'
  | 'bytes4[]'
  | 'bytes32[]'

/**
 * Field validation rules
 */
export interface FieldValidation {
  /** Minimum value (for numbers) */
  min?: string

  /** Maximum value (for numbers) */
  max?: string

  /** Regex pattern (for strings) */
  pattern?: string

  /** Custom validation message */
  message?: string
}

/**
 * Installed module on a Smart Account
 */
export interface InstalledModule {
  /** Module contract address */
  address: Address

  /** Module type */
  type: ModuleType

  /** Module metadata */
  metadata: ModuleMetadata

  /** Installation configuration (encoded) */
  initData: Hex

  /** Decoded configuration for display */
  config?: Record<string, unknown>

  /** Installation status */
  status: ModuleStatus

  /** Installation timestamp */
  installedAt?: number

  /** Installation transaction hash */
  installTxHash?: Hex
}

/**
 * Module installation request
 */
export interface ModuleInstallRequest {
  /** Module address to install */
  moduleAddress: Address

  /** Module type */
  moduleType: ModuleType

  /** Encoded initialization data */
  initData: Hex
}

/**
 * Module uninstallation request
 */
export interface ModuleUninstallRequest {
  /** Module address to uninstall */
  moduleAddress: Address

  /** Module type */
  moduleType: ModuleType

  /** Encoded de-initialization data */
  deInitData: Hex
}

// ============================================================================
// Validator-Specific Types
// ============================================================================

/**
 * ECDSA Validator configuration
 */
export interface ECDSAValidatorConfig {
  /** Owner address that can sign */
  owner: Address
}

/**
 * WebAuthn Validator configuration
 */
export interface WebAuthnValidatorConfig {
  /** WebAuthn credential public key X coordinate */
  pubKeyX: bigint

  /** WebAuthn credential public key Y coordinate */
  pubKeyY: bigint

  /** Credential ID */
  credentialId: Hex
}

/**
 * MultiSig Validator configuration
 */
export interface MultiSigValidatorConfig {
  /** List of signer addresses */
  signers: Address[]

  /** Required signatures threshold */
  threshold: number
}

// ============================================================================
// Executor-Specific Types
// ============================================================================

/**
 * Session Key configuration
 */
export interface SessionKeyConfig {
  /** Session key address */
  sessionKey: Address

  /** Allowed target contracts */
  allowedTargets: Address[]

  /** Allowed function selectors */
  allowedSelectors: Hex[]

  /** Maximum value per transaction */
  maxValuePerTx: bigint

  /** Expiration timestamp */
  validUntil: number

  /** Start timestamp */
  validAfter: number
}

/**
 * Recurring Payment configuration
 */
export interface RecurringPaymentConfig {
  /** Recipient address */
  recipient: Address

  /** Token address (0x0 for native) */
  token: Address

  /** Payment amount */
  amount: bigint

  /** Interval in seconds */
  interval: number

  /** Maximum number of payments (0 for unlimited) */
  maxPayments: number
}

// ============================================================================
// Hook-Specific Types
// ============================================================================

/**
 * Spending Limit Hook configuration
 */
export interface SpendingLimitHookConfig {
  /** Token address (0x0 for native) */
  token: Address

  /** Spending limit amount */
  limit: bigint

  /** Reset period in seconds */
  period: number
}

/**
 * Audit Hook configuration
 */
export interface AuditHookConfig {
  /** Events to log */
  logEvents: ('transfer' | 'approval' | 'call')[]

  /** Whether to log calldata */
  logCalldata: boolean
}

// ============================================================================
// Type Guards
// ============================================================================

/**
 * Check if module type is Validator
 */
export function isValidator(type: ModuleType): boolean {
  return type === MODULE_TYPE.VALIDATOR
}

/**
 * Check if module type is Executor
 */
export function isExecutor(type: ModuleType): boolean {
  return type === MODULE_TYPE.EXECUTOR
}

/**
 * Check if module type is Hook
 */
export function isHook(type: ModuleType): boolean {
  return type === MODULE_TYPE.HOOK
}

/**
 * Check if module type is Fallback
 */
export function isFallback(type: ModuleType): boolean {
  return type === MODULE_TYPE.FALLBACK
}

/**
 * Check if module type is Policy
 */
export function isPolicy(type: ModuleType): boolean {
  return type === MODULE_TYPE.POLICY
}

/**
 * Check if module type is Signer
 */
export function isSigner(type: ModuleType): boolean {
  return type === MODULE_TYPE.SIGNER
}

/**
 * Get module type name
 */
export function getModuleTypeName(type: ModuleType): string {
  return MODULE_TYPE_NAMES.get(type) ?? 'Unknown'
}

// ============================================================================
// Encoding/Decoding Types
// ============================================================================

/**
 * Module ABI for encoding/decoding
 */
export interface ModuleABI {
  /** Install function signature */
  installSelector: Hex

  /** Uninstall function signature */
  uninstallSelector: Hex

  /** Init data encoding types */
  initDataTypes: SolidityType[]

  /** De-init data encoding types */
  deInitDataTypes: SolidityType[]
}
